// Copyright 2015, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

syntax = "proto3";

package assignment2;

import "google/protobuf/timestamp.proto";

// The messenger service definition.
service FBChatServer {
  rpc Login (ClientRequest) returns (ServerReply) {}
  rpc Connect (ClientRequest) returns (ServerReply) {}
  rpc List (ClientRequest) returns (ShowList) {}
  rpc Join (ClientRequest) returns (ServerReply) {}
  rpc Leave (ClientRequest) returns (ServerReply) {}
  rpc Chat (stream Post) returns (stream Post) {}
  rpc Alive (ClientRequest) returns (ServerReply) {}
  rpc Check (ClientRequest) returns (ServerReply) {}
}

service RegisterServer{
	rpc ServerRegister(ServerReply) returns (ServerReply){}
    rpc Join(DataSync) returns (ServerReply){}
    rpc Login(DataSync) returns (ServerReply){}
    rpc Leave(DataSync) returns (ServerReply){}
    rpc updateTimeLine(DataSync) returns (ServerReply){}
    rpc msgForward(DataSync) returns (ServerReply){}
}

message ClientRequest {
  string username = 1;
  repeated string requestinfo = 2;
}

message ShowList {
  repeated string all_clients = 1;
  repeated string joined_clients = 2;
}

message ServerReply {
  string message = 1;
   int32 id = 2;
}

message DataSync {
    string message = 1;
    string username = 2;
    string targetname = 3;
    string servername = 4;
}

message Post {
  string username = 1;
  string content = 2;
  google.protobuf.Timestamp timestamp = 3;
}

